src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs

```diff
@@ -68,11 +74,22 @@ public sealed override async Task WriteResponseBodyAsync(OutputFormatterWriteCon
 
         var httpContext = context.HttpContext;
 
-        // context.ObjectType reflects the declared model type when specified.
-        // For polymorphic scenarios where the user declares a return type, but returns a derived type,
-        // we want to serialize all the properties on the derived type. This keeps parity with
-        // the behavior you get when the user does not declare the return type and with Json.Net at least at the top level.
-        var objectType = context.Object?.GetType() ?? context.ObjectType ?? typeof(object);
+        // Maybe we could use the jsontypeinfo overload but we need the untyped,
+        // waiting for https://github.com/dotnet/runtime/issues/77051
+
+        var declaredType = context.ObjectType;
+        var runtimeType = context.Object?.GetType();
+        var objectType = runtimeType ?? declaredType ?? typeof(object);
+
+        if (declaredType is not null &&
+            runtimeType != declaredType &&
+            SerializerOptions.TypeInfoResolver != null &&
+            SerializerOptions.GetTypeInfo(declaredType).PolymorphismOptions is not null)
```

Comment:
`JsonTypeInfo` is being looked up twice from the `JsonSerializerOptions` cache, both here and the `SerializeAsync` method in line 99. You could just reuse the metadata token using the new untyped JsonTypeInfo APIs.

Reply:
The dependency update is blocked but as soon as we have the repo updated I will change it.